Update "Data Types" Page

Savio-Sou opened the pull request:
Closes https://github.com/noir-lang/docs/issues/54, Closes https://github.com/noir-lang/docs/issues/43.

# "Data Types" Page Changes
- Add a "Boolean Type" section
- Move Constants & Globals sections out to the "Mutability" page
- Simplify the Tuple example demonstrating access via destructuring
- Modify Struct examples to use less technical terms
- Improve reading / comprehension flow
    - Add a Private & Public Types section
    - Swap the orders to introduce Tuples first then Structs
    - Minor readability edits

# "Hello, World" Page Changes
- Add a reference to the "Data Types" page for more info on private & public values
- Minor fixes

Savio-Sou committed:
Review "Data Types" Page

Savio-Sou committed:
Merge branch 'master' into data-types

Globallager commented:
# Preview Changes
1. `git clone https://github.com/noir-lang/book.git`
2. `cd book`
3. `git fetch origin pull/42/head:pr/42`
4. `git checkout pr/42`
5. `mdbook serve --open`

Savio-Sou committed:
Minor Grammar Fix

vezenovm reviewed (approved):
These changes look good to me

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+# Data Types
 
-Noir has several varieties of data types: the primitive `Field` type, integer types, `bool`, arrays, structs, and tuple types.
+Every value in Noir is specified with a certain data type, which informs Noir tools how to work with that data when processing Noir programs.
```
Lets reword this. "is specified with" implies types are required to be specified for all values rather than being inferred. We can say "Every value in noir has a type which determines which operations are valid for it." The point on tooling seems unnecessary especially since the only tooling that looks at types currently is the compiler itself.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
 
-Although each value in a constraint system is fundamentally a field element, we add a layer of abstraction over this;
-each value can be _concealed_ (a private type) or _revealed_ (a public type).
+All values in Noir are fundamentally `Field` elements. For a more approachable developing experience, abstractions are added on top to introduce different data types in Noir.
```
"fundamentally composed of `Field` elements` would be more accurate since some values like arrays may contain multiple.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
-A **private value** is known only to the Prover, while, a **public value** is known by the Prover and Verifier.
-All public and private values are primitive types, though individual
-fields of a struct may also be marked private or public.
+Noir has two category of data types: primitive types (e.g. `Field`, integer, `bool`) and compound types (e.g. array, tuple, struct). Each value can either be private or public.
```
lets change "integer" to "integers" or "signed/unsigned integers". The compound types listed should also be plural - "arrays, tuples, and structs"

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
-Any of these types can be private (the default) or public (via a `pub` modifier).
-
-Note that private types are also referred to as witnesses.
+A **private value** is known only to the Prover, while a **public value** is known by both the Prover and Verifier. All public and private values are primitive types, though individual fields of compound types may also be marked private or public.
```
We don't allow marking something explicitly private since it is the default. It'd be more accurate to say fields are also private by default but can be explicitly marked public.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+Public data types are treated no differently to private types apart from the fact that their values will be revealed in proofs generated. Simply changing the value of a public type will not change the circuit (where the same goes for changing values of private types as well).
 
-`x`, `y` and `z` are witness types (private `Field`). Using the `let` keyword we derived a new witness `z` which is constrained to be equal to `x + y`.
+_Private_ are also referred to as _witnesses_ sometimes.
```
Should be "Private values"

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
-#### Integer Type
+> **Note:** The terms private and public when applied to a type (e.g. `pub Field`) have a different meaning than when applied to a function (e.g. `pub fn foo() {}`).
+>
+> The latter is a visibility modifier familiar to those coming from other languages, while the former can be thought of as a visibility modifier for the Verifier to interpret.
```
Can we re-order this so that the former is explained first? I think it'd be less confusing. If we want to introduce the more familiar function visibility first we should re-order the first sentence on line 19 then.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
 
-`x` and `y` are both private types (witnesses), however `y` is an integer type.
-If `y` exceeds the range \\([0,2^{24}-1]\\) then any proof created will output false by the verifier.
+In this example, the first input (i.e. `x`) is private while the second input (i.e. `y`) and the return value (i.e. `x + y`) are public.
```
Too much use of (i.e) here. Would be more natural to remove it:
> In this example, `x` is private while `y` and the return value are public. Note that visibility is handled per-variable, so it is perfectly valid to have one input that is private and another that is public.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+In this example, the first input (i.e. `x`) is private while the second input (i.e. `y`) and the return value (i.e. `x + y`) are public.
 
-> **Note:** The Aztec backend supports even and odd sized integer types, so while using the Aztec backend, even sized integer types such as u32, u48, as well as odd sized integer types such as u5, u3 will produce proofs.
+> **Note:** As of this moment, public types can only be declared through parameters on `main`.
```
Lets remove "As of this moment." All information in the docs is as of this moment and subject to change

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+The `Field` type corresponds to the native field type of the proving backend. For example, this would be a (roughly) 256-bit integer when using the default TurboPlonk backend.
 
-Note that variables declared as mutable may not be constants:
+Fields are simply defined with the keyword `Field`:
```
Fields are not defined with the Field keyword, they are defined with `let` statements like other values. I think we can remove this sentence. Perhaps replacing it with a brief explanation of ways in which we can use fields. E.g. "Fields support integer arithmetic and are often used as the default numeric type in Noir" or similar.

jfecher commented on src/language_concepts/data_types.md:
````diff
@@ -1,206 +1,192 @@
 ```
 
-#### Globals
+`x`, `y` and `z` are all private `Field` in this example. Using the `let` keyword we derived a new private value `z` constrained to be equal to `x + y`.
````
It is a bit awkward with the code quoting, but we could say "are all private `Field`s" to fix the missing plural.
"derived" -> "defined"

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+`x`, `y` and `z` are all private `Field` in this example. Using the `let` keyword we derived a new private value `z` constrained to be equal to `x + y`.
 
-Noir also supports global variables. However, they must be compile-time variables. If `comptime` is not explicitly written in the type annotation the compiler will implicitly specify the declaration as compile-time. They can then be used like any other compile-time variable inside functions. The global type can also be inferred by the compiler entirely. Globals can also be used to specify array a ...
+If efficiency is of priority, `Field` should be used as the default type for solving problems. Using smaller integer types (e.g. `u64`) incurs extra range constraints.
```
Oh perhaps this could be used/mentioned/rearranged with the removed blurb on line 45.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+### The Boolean Type
 
-A public type is a value that may change per circuit instance. Unlike Constants, changing a public value will not change the circuit.
+A boolean type in Noir has two possible values: `true` and `false`, and is specified using `bool`. For example:
```
_The_ `bool` type in Noir has two possible values: `true` and `false`.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
-
+fn main() {
+    let t = true; // type implicitly interpreted
+    let f: bool = false; // type explicit annotated
```
We should remove these comments. This is an odd place to start mentioning type inference when we've already used it in the examples above.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+> **Note:** When returning a boolean value, it will show up as a value of 1 for `true` and 0 for `false` in _Verifier.toml_.
 
-> **Note:** This behaviour and type will change in future releases, to catch a linearity bug in user code.
+The main usage of the boolean type is in conditionals (e.g. `if` expressions, `constrain`). More about conditionals are covered in the [Control Flow](./control_flow.md) and [Constrain Statement](./constrain.md) sections.
```
This reads a bit awkwardly. Lets rephrase the first sentence to: "The boolean type is most commonly used in conditionals like `if` expressions and `constrain` statements."

jfecher commented on src/language_concepts/data_types.md:
````diff
@@ -1,206 +1,192 @@
 ```
 
-Note that currently Noir only supports arrays with integer or `Field` elements.
+All elements in an array must be of the same type (i.e. homogeneous). That is, `Field` values cannot be grouped together with integer values in one array for example.
````
Since this is targetted towards beginners, lets make the example a bit less generic and use "... together with `u8` values" rather than "integer values" since that may lead to more questions for users confused with the difference between Fields and integers.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+All elements in an array must be of the same type (i.e. homogeneous). That is, `Field` values cannot be grouped together with integer values in one array for example.
 
-> **Example:** An array of `Field` types cannot be grouped together with an array of integer types.
+Noir currently supports arrays with `Field` or integer elements only.
```
I believe we also support boolean arrays (? citation needed). So perhaps this should be "arrays with primitive element types only"

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+A compound type groups together multiple values into one type. Elements within a compound type can be private or public.
 
-#### Arrays
+### The Array Type
```
This section should show what the array type looks like: `[Field; 3]` for example.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
-#### Tuples
+### Structs
+
+A struct is used for grouping multiple related values of different types like a tuple, with the added ability to name the pieces of data (also referred to as _fields_).
```
This reads a bit awkwardly with the ", with" and no conjunction. Perhaps we can rephrase it:
"A struct also allows for grouping multiple values of different types. Unlike tuples, we can also name each field. Note that the usage of "field" here is referring to each element of the struct and is unrelated to the `Field` type."

So we can merge the note below into the paragraph itself. I've also removed the mention of Rust since the phrase "a struct's fields" does not originate with rust nor was it popularized by it. If we want to keep the reasoning though we could mention that this follows convention established by other programming languages.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+
+> **Note:** The naming of _fields_ here is unrelated to Noir's `Field` type. It simply mimics Rust's convention.
+
+When defining a struct, you will name the struct and every fields within as `key: type` pairs. A struct can hold fields of different types:
```
Defining a struct requires giving it a name and listing each field within as `key: type` pairs.


I don't think we should include "A struct can hold fields of different types" since we already mentioned that above.

jfecher commented on src/language_concepts/data_types.md:
````diff
@@ -1,206 +1,192 @@
+}
+```
+
+An instance of struct can then be created with concrete values in `key: value` pairs. Struct fields are accessible using their given names, instead of by their order like in arrays and tuples:
````
We can rephrase this as:

"When initializing a struct value, its fields can be listed in any order."

Globallager replied:
Plan was to highlight accessing struct fields with names instead of e.g. Animal.0 or Animal[0], but simultaneously showcasing field listing in any order does sound like a good idea.

Revised in upcoming commit.

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -1,206 +1,192 @@
+    let dog = Animal {
+        hands: 0,
+        legs,
+        eyes: 2, // type implicitly interpreted
```
Comment can be removed

jfecher commented on src/language_concepts/mutability.md:
```diff
@@ -43,6 +43,63 @@ fn helper(mut x: i32) {
 
+### Constants
+
+A constant type is a value that does not change per circuit instance. This is different to a witness which changes per proof. If a constant type that is being used in your program is changed, then your circuit will also change.
```
I think this whole section should be reworded, we've been avoiding the use of `const` internally since it can mean several different things. This section seems to introduce `comptime` values so I think it'd be less confusing to call them `comptime` values. We could say something along the lines of "comptime values are values that are known at compile-time" or "evaluated to a constant value at compile-time" or similar.

Globallager replied:
Thanks for the insights.

The "Constants" and "Globals" sections were simply cut-pasted to this page and are intended to be updated with a separate PR. I'll keep the suggestion in mind when I get to it 🙌

jfecher commented on src/language_concepts/mutability.md:
```diff
@@ -43,6 +43,63 @@ fn helper(mut x: i32) {
+fn main() {
+    let a: comptime Field = 5;
+
+    // `comptime Field` can also be inferred:
```
Unlike the other sections, I'm fine with this note on inference here as "comptime" is likely something that will be unfamiliar to most users and they may incorrectly assume it is not part of the type and thus cannot be inferred otherwise.

Globallager replied:
Thanks, will keep the suggestion in mind 🙌

Savio-Sou committed:
Jake's Suggestions + Edits
- Revise based on Jake's suggestions
- Add array example illustrating array element access
- Minor edits

Globallager commented:
Thanks for spending the time to review it @jfecher & @vezenovm. New commit pushed w/ screenshot attached for easy preview (open in new tab + zoom in):
![Data Types](https://user-images.githubusercontent.com/72797635/215045472-bcbad28c-abdd-4d60-a4c3-6ecd59d18b14.png)

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -30,33 +30,35 @@ fn main(x : Field, y : pub Field) -> pub Field {
+`x`, `y` and `z` are all private fields in this example. Using the `let` keyword we defined a new private value `z` constrained to be equal to `x + y`.
 
-If efficiency is of priority, `Field` should be used as the default type for solving problems. Using smaller integer types (e.g. `u64`) incurs extra range constraints.
+If proving efficiency is of priority, fields should be used as a default for solving problems. Smaller integer types (e.g. `u64`) come with overflow protections at the cost of extra range constraints.
```
This seems like it suggests integer types will fail when they overflow, but the current behavior is to wrap around instead. It sounds like we want to look into changing this default so I'm not sure which we should mention here. Perhaps we should only mention that they incur extra range constraints.

Globallager replied:
Thanks for the correction 🙏

jfecher commented on src/language_concepts/data_types.md:
```diff
@@ -66,42 +68,54 @@ fn main(x : Field, y : u32) {
 ### The Array Type
 
-An array is one way of grouping together values into one compound type:
+An array is one way of grouping together values into one compound type. It can be implicitly inferred or explicitly defined with `[<Type>; <Size>]`:
```
"Defined" is for defining a name to be equal to a value. Since there is no value here, it'd be more accurate to say "Array types can be be inferred or explicilty specified via the syntax `[<Type>; <Size>]`"

jfecher commented on src/language_concepts/data_types.md:
````diff
@@ -66,42 +68,54 @@ fn main(x : Field, y : u32) {
 ```
 
-All elements in an array must be of the same type (i.e. homogeneous). That is, `Field` values cannot be grouped together with integer values in one array for example.
+Here, both `my_arr` and `your_arr` are instantiated as an array of `Field` type with `2` elements.
````
Grammar - should be "as an array containing two `Field` elements"

jfecher commented on src/language_concepts/data_types.md:
````diff
@@ -152,16 +166,16 @@ struct Animal {
 ```
 
-An instance of struct can then be created with concrete values in `key: value` pairs. Struct fields are accessible using their given names, instead of by their order like in arrays and tuples:
+An instance of struct can then be created with actual values in `<Key>: <Value>` pairs of any order. Struct fields are accessible using their given names:
````
"An instance of a struct"
"pairs in any order"

jfecher reviewed (commented):
A few more small grammar items

Savio-Sou committed:
Jake's Followup Suggestions

Globallager commented:
> A few more small grammar items

Thanks for the followup review @jfecher!

jfecher reviewed (approved):
Thank you!

Savio-Sou committed:
Merge branch 'noir-lang:master' into data-types
